amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,422 @@
+//! Contains the `Error` type and company as used by Amethyst.
+//!
+//! **Note:** This type is not intended to be used outside of Amethyst.
+//! If you are integrating a crate from amethyst to use this, it is recommended that you treat this
+//! type as an opaque [`std::error::Error`].
+//!
+//! [`std::error::Error`]: https://doc.rust-lang.org/std/error/trait.Error.html
+
+// Parts copied from failure:
+// https://github.com/rust-lang-nursery/failure
+
+#![warn(missing_docs, rust_2018_idioms, rust_2018_compatibility)]
+
+#[cfg(all(feature = "backtrace", feature = "std"))]
+extern crate backtrace;
+
+use self::internal::Backtrace;
+use std::borrow::Cow;
+use std::error;
+use std::fmt;
+use std::result;
+
+/// Internal parts of `Error`.
+#[derive(Debug)]
+struct Inner<T: ?Sized> {
+    source: Option<Box<Error>>,
+    #[cfg(all(feature = "backtrace", feature = "std"))]
+    backtrace: Option<Backtrace>,
+    error: T,
+}
+
+/// The error type used by Amethyst.
+///
+/// Wraps error diagnostics like messages and other errors, and keeps track of causal chains and
+/// backtraces.
+pub struct Error {
+    inner: Box<Inner<dyn error::Error + Send + Sync>>,
+}
+
+impl Error {
+    /// Default constructor for our error types.
+    ///
+    /// Wraps anything that is an error in a box.
+    pub fn new<E>(error: E) -> Self
+    where
+        E: 'static + error::Error + Send + Sync,
+    {
+        Self {
+            inner: Box::new(Inner {
+                source: None,
+                #[cfg(all(feature = "backtrace", feature = "std"))]
+                backtrace: self::internal::bt::new(),
```

Comment:
So each time I wrap it collects brand new backtrace.
```
fn bar() -> Result<(), Error> {
  foo().with_context(|_| Error::from_string("foo failed in `bar`"))
}
```
Now `foo` returns an error with backtrace and `bar` captures one more redundant backtrace.

Reply:
So is your proposal is to only capture the backtrace for the root error? How would we do this?